### DISCUSSION TITLE
Can't get the rule's command BREAK to stop furhter rules evaluation

### DISCUSSION THREAD

**Discussion by nonix** [arendst/Tasmota#13229] [Q&A] [answered]
Hi guys, can anyone help?
I have a rule:
_rule3 on AM2301#Temperature == null do break 
          on AM2301#Temperature < %mem1% do power 1 endon
          on AM2301#Temperature > %mem2% do power 0 endon_

Which seems to trigger just fine, but the break does not break 

After switching the rule3 ON it triggers both (all) conditions

_14:47:51.153 CMD: rule3 on
14:47:51.162 MQT: stat/heater/RESULT = {"Rule3":{"State":"ON","Once":"ON","StopOnError":"OFF","Length":134,"Free":377,"Rules":"on AM2301#Temperature == null do break on AM2301#Temperature < %mem1% do power 1 endon on AM2301#Temperature > %mem2% do power 0 endon"}}
14:47:51.206 RUL: AM2301#TEMPERATURE == NULL performs ""
14:47:51.928 RUL: AM2301#TEMPERATURE < %MEM1% performs "power 1"
14:47:51.938 MQT: stat/heater/RESULT = {"POWER":"ON"}
14:47:51.942 MQT: stat/heater/POWER = ON_

This is the tele message from the sensor:
_14:52:24.461 MQT: tele/heater/SENSOR = {"Time":"2021-09-28T14:52:24","AM2301":{"Temperature":null,"Humidity":null,"DewPoint":null},"TempUnit":"C"}_

FYI:
 MEM1 3
 MEM2 5

I am running Tasmota V9.5.0 on ESP8266 SonoOff TH.

What am I doing wrong?
Thank you,
 N.

**Comment by ascillato** [comment, marked as answer]
Hi, please try the following that don't have extra spaces:

```js
Rule3
   ON AM2301#Temperature=null DO Delay BREAK
   ON AM2301#Temperature<%mem1% DO Power 1 BREAK
   ON AM2301#Temperature>%mem2% DO Power 0 ENDON
```

All together in one line so as to copy and paste:
```js
Rule3 ON AM2301#Temperature=null DO Delay BREAK ON AM2301#Temperature<%mem1% DO Power 1 BREAK ON AM2301#Temperature>%mem2% DO Power 0 ENDON
```

**Reply by nonix** [reply]
Hi, thank you very much for helping.

I did as you have suggest, but seems did not break after first match, though it did triggered.

```
16:08:40.772 CMD: Rule3 ON AM2301#Temperature=null DO Delay BREAK ON AM2301#Temperature<%mem1% DO Power 1 BREAK ON AM2301#Temperature>%mem2% DO Power 0 ENDON
16:08:40.779 RUL: Stored uncompressed, would compress from 133 to 75 (-44%)
16:08:40.784 MQT: stat/heater/RESULT = {"Rule3":{"State":"ON","Once":"ON","StopOnError":"OFF","Length":133,"Free":378,"Rules":"ON AM2301#Temperature=null DO Delay BREAK ON AM2301#Temperature<%mem1% DO Power 1 BREAK ON AM2301#Temperature>%mem2% DO Power 0 ENDON"}}
16:08:41.124 RUL: AM2301#TEMPERATURE=NULL performs "Delay"
16:08:41.134 MQT: stat/heater/RESULT = {"Delay":0}
16:08:41.831 RUL: AM2301#TEMPERATURE<%MEM1% performs "Power 1"
16:08:41.839 MQT: stat/heater/RESULT = {"POWER":"ON"}
16:08:41.842 MQT: stat/heater/POWER = ON
16:08:44.447 MQT: tele/heater/STATE = {"Time":"2021-09-28T16:08:44","Uptime":"0T01:35:29","UptimeSec":5729,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"FB-CDM","BSSId":"5C:49:79:23:E4:1B","Channel":1,"Mode":"11n","RSSI":100,"Signal":-43,"LinkCount":1,"Downtime":"0T00:00:03"}}

```


I have also attached _status 0_  as a reference.
```
16:12:34.916 MQT: stat/heater/STATUS = {"Status":{"Module":4,"DeviceName":"Heater Control","FriendlyName":["heater"],"Topic":"heater","ButtonTopic":"0","Power":1,"PowerOnState":0,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":0,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
16:12:34.935 MQT: stat/heater/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Power On","Uptime":"0T01:39:19","StartupUTC":"2021-09-28T12:33:15","Sleep":50,"CfgHolder":4617,"BootCount":593,"BCResetTime":"2021-06-14T10:21:15","SaveCount":950,"SaveAddress":"F8000"}}
16:12:34.950 MQT: stat/heater/STATUS2 = {"StatusFWR":{"Version":"9.5.0(tasmota)","BuildDateTime":"2021-06-17T08:26:35","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"349/699"}}
16:12:34.967 MQT: stat/heater/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["FB-CDM",""],"TelePeriod":20,"Resolution":"558180C0","SetOption":["00008008","2805C8000100060000005A0A000000000000","00000080","00006000","00000000"]}}
16:12:34.988 MQT: stat/heater/STATUS4 = {"StatusMEM":{"ProgramSize":607,"Free":396,"Heap":26,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"146085","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
16:12:35.002 MQT: stat/heater/STATUS5 = {"StatusNET":{"Hostname":"Heater","IPAddress":"192.168.188.61","Gateway":"192.168.188.1","Subnetmask":"255.255.255.0","DNSServer":"192.168.188.1","Mac":"84:CC:A8:94:78:EF","Webserver":2,"WifiConfig":0,"WifiPower":17.0}}
16:12:35.016 MQT: stat/heater/STATUS6 = {"StatusMQT":{"MqttHost":"icontroller1.fritz.box","MqttPort":1883,"MqttClientMask":"heater","MqttClient":"heater","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
16:12:35.033 MQT: stat/heater/STATUS7 = {"StatusTIM":{"UTC":"2021-09-28T14:12:35","Local":"2021-09-28T16:12:35","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":99,"Sunrise":"07:45","Sunset":"19:35"}}
16:12:35.050 MQT: stat/heater/STATUS10 = {"StatusSNS":{"Time":"2021-09-28T16:12:35","AM2301":{"Temperature":null,"Humidity":null,"DewPoint":null},"TempUnit":"C"}}
16:12:35.066 MQT: stat/heater/STATUS11 = {"StatusSTS":{"Time":"2021-09-28T16:12:35","Uptime":"0T01:39:20","UptimeSec":5960,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"FB-CDM","BSSId":"5C:49:79:23:E4:1B","Channel":1,"Mode":"11n","RSSI":100,"Signal":-41,"LinkCount":1,"Downtime":"0T00:00:03"}}}
16:12:44.431 MQT: tele/heater/STATE = {"Time":"2021-09-28T16:12:44","Uptime":"0T01:39:29","UptimeSec":5969,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"FB-CDM","BSSId":"5C:49:79:23:E4:1B","Channel":1,"Mode":"11n","RSSI":100,"Signal":-42,"LinkCount":1,"Downtime":"0T00:00:03"}}
16:12:44.439 MQT: tele/heater/SENSOR = {"Time":"2021-09-28T16:12:44","AM2301":{"Temperature":null,"Humidity":null,"DewPoint":null},"TempUnit":"C"}

```

Thank you,
 N.

**Reply by sfromis** [reply]
The time difference of 0.7 seconds from the trigger for `null` firing to the less-than rule firing looks like it may be the next cycle triggering rule number 2. And when you also use once, the first rule may be deactivated as long as the condition is still true, opening up the field to next rule.

**Reply by ascillato** [reply]
> The time difference of 0.7 seconds from the trigger for null firing to the less-than rule firing looks like it may be the next cycle triggering rule number 2. And when you also use once, the first rule may be deactivated as long as the condition is still true, opening up the field to next rule.

Interesting...

@nonix

Please, can you try to disable one shot by command `rule3 4` and try again?

**Comment by nonix** [comment]
Hi guys, it seems that did the trick, but side effect (as expected) it fires the rule at constant:
```
17:13:51.952 RUL: AM2301#TEMPERATURE=NULL performs "Delay"
17:13:51.958 MQT: stat/heater/RESULT = {"Delay":0}
17:13:52.553 RUL: AM2301#TEMPERATURE=NULL performs "Delay"
17:13:52.559 MQT: stat/heater/RESULT = {"Delay":0}
17:13:53.332 RUL: AM2301#TEMPERATURE=NULL performs "Delay"
17:13:53.339 MQT: stat/heater/RESULT = {"Delay":0}
17:13:53.932 RUL: AM2301#TEMPERATURE=NULL performs "Delay"
17:13:53.939 MQT: stat/heater/RESULT = {"Delay":0}
```


'm going to try to reset all to fw default values and re-try ...

**Reply by ascillato** [reply]
You can then use `tele-AM2301#TEMPERATURE=NULL` as a trigger to limit the checks to only the teleperiod time

**Comment by nonix** [comment]
OK, that did the trick, though IMHO this should be logged as a bug. Would you agree?

FYI, the complete reset and re-config (rule with AM2301 only) behaves just the same way.

**Reply by sfromis** [reply]
What's the bug? Meaning how is it not operating according to documentation?

Details of how rule execution works can be tricky and somewhat confusing at times, but that does not make it a bug when things did not do what you expected. Maybe something could be improved, but that might break existing rules.

**Reply by nonix** [reply]
@sfromis , Logging a bug is good think not bad. It says people like it and do use it. 
As **meingraham** pointed out, I have only followed [Features/Rules/Syntax](https://tasmota.github.io/docs/Rules/) paragraph:

> BREAK - marks the end of a rule. BREAK will stop the execution of the remaining rules that follow this rule within the rule set. If a rule that ends with BREAK is triggered, the following rules in that rule set will not be executed. This allows the rules to somewhat simulate an "IF/ELSE" statement.

**Comment by meingraham** [comment]
I do think that the behavior of `BREAK` has changed sometime since early v8.x. Previously, `BREAK` along with a one-shot rule set worked as documented. Lately I've seen a few `BREAK` related posts where it no longer acts as documented including the cookbook examples.

**Comment by nonix** [comment]
OK, now I understnad the woraround.

And here is my 5¢ perhpas the issue could be solved by adjusting behaviour of NULL, where a **null value should not evaluate to true (null<1)** it should be false.

True, this does not fix the BREAK issue, but in my case the BREAK would not be needed, as the BREAK was my proposed woraround for incorrect evaluation of null.

Thank you to all for helping, clarifying and mainting this great open source firmware.
N.
